airlock/business_logic.py

```diff
@@ -1669,6 +1708,48 @@ def reset_review_file(
 
         self._dal.reset_review_file(release_request.id, relpath, user.username, audit)
 
+    def review_request(self, release_request: ReleaseRequest, user: User):
+        """
+        Complete a review
+
+        Marking the request as either PARTIALLY_REVIEWED or REVIEWED, depending on whether this is the first or second review.
+        """
+        if self.STATUS_OWNERS[release_request.status] != RequestStatusOwner.REVIEWER:
+            raise self.RequestPermissionDenied(
+                f"Cannot review request in state {release_request.status.name}"
+            )
+
+        if user.username == release_request.author:
+            raise self.RequestPermissionDenied("You cannot review your own request")
+
+        if not user.output_checker:
+            raise self.RequestPermissionDenied(
+                "Only an output checker can review a request"
+            )
+
+        if not release_request.all_files_reviewed_by_reviewer(user):
+            raise self.RequestReviewDenied(
+                "You must review all files to complete your review"
+            )
+
+        if user.username in release_request.completed_reviews:
+            raise self.RequestReviewDenied(
+                "You have already completed your review of this request"
+            )
+
+        self._dal.record_review(release_request.id, user.username)
+
+        release_request = self.get_release_request(release_request.id, user)
+        n_reviews = len(release_request.completed_reviews)
+
+        # this method is called twice, by different users. It advances the
+        # state differently depending on whether its the 1st or 2nd review to
+        # be completed.
+        if n_reviews == 1:
+            self.set_status(release_request, RequestStatus.PARTIALLY_REVIEWED, user)
+        elif n_reviews == 2:
+            self.set_status(release_request, RequestStatus.REVIEWED, user)
```

Comment:
this block is screaming 'race condition' to me, but we're hardly google-scale, maybe it's fine?

Reply:
(to properly fix it we'd need to push some business logic out of the BLL & into the DAL??)

Reply:
I guess it is possible; if R1 completes their review and status is set to partially reviewed, and at the same time R2 completes their review and by the time it tries to set status to partially reviewed, it's already partially reviewed, we get an invalid state transition error. Given that the states for partially reviewed/reviewed can be determined by the current state of the completed reviews, we could check those in the event of an invalid state transition here, and set it to the right one.

Reply:
oh yeah, that's a good point, we could gracefully handle it if we want to. let's not worry about it for now, there's already a lot going on.

Reply:
I've added some handling of the race condition for 2 checkers approving simultaneously, although at the same time it occurred to me that when a request is fully approved it can be returned/released/rejected, so potentially a 3rd reviewer could attempt to complete a review at the same time that another review returns/releases/rejects and puts it into an invalid state for reviewing.  In that case, it'll just re-raise the exception - which I think is OK, given how unlikely it is to happen.